students/kegan/session03/mailroom.py

```diff
@@ -0,0 +1,301 @@
+    update_widths(headers)
+    data = []
+    # obtain data and determine ideal column lengths before making report
+    for donor in sorted(DONORS, key=by_donation, reverse=True):
+        donations = DONORS[donor]
+        line = [donor, sum(donations), len(donations), average(donations)]
+        update_widths(line)
+        data.append(line)
+    report = []
+    report.append(stringify(headers, '| '))
+    line_width = sum(COLUMN_WIDTHS) + len(COLUMN_WIDTHS) * 2 + 1
+    report.append(horizontal_line(line_width))
+    for row in data:
+        report.append(stringify(row, ' '))
+    print('\n' + '\n'.join(report) + '\n')
+
+
+def update_widths(line):
+    """ Updates minimum column widths based on
+    the length of items (as strings) in given line.
+    Args:
+        line (list) : list of values
+    """
+    for index in range(len(line)):
+        item = line[index]
+        # dollar amounts are slightly longer in final report
+        if type(item) is float:
+            item = dollar(item, report=True)
+        item = str(item)
+        if len(item) > COLUMN_WIDTHS[index]:
+            COLUMN_WIDTHS[index] = len(item)
+
+
+def write_thank_yous():
+    """ Writes thank yous to all donors in individual
+    files in a ThankYous folder in the program's cwd."""
+    import os
+    print('Writing new thank yous to all donors...')
+    for donor in DONORS:
+        thankyou = get_thank_you(donor, DONORS[donor])
+        if not os.path.exists('ThankYous'):
+            os.mkdir('ThankYous')
+        with open(os.path.join('ThankYous', donor + '.txt'), 'w') as f:
+            f.write(thankyou)
+    print('Thank yous written to\n{}'.format(
+        os.path.join(os.getcwd(), 'ThankYous')))
+
+
+def by_donation(donor):
+    """ Returns the sum of the donations for given donor.
+    Args:
+        donor (str) : donor name
+    Returns:
+        float : sum of all donations for given donor
+    """
+    return sum(DONORS[donor])
+
+
+def average(donations):
+    """ Returns average of given donations rounded to 2 decimals.
+    Args:
+        donations (list of floats) : list of donations
+    Returns:
+        float : average of donations
+    """
+    avg = sum(donations) / len(donations)
+    return round(avg, 2)
+
+
+def dollar(amount, report=False):
+    """ Returns amount in a format suitable for showing
+    dollar amounts, e.g. 181.98 or 211. Removes cents
+    from any whole dollar amount.
+     Args:
+        amount (float) : amount to turn into dollar format
+        report (True) : True returns report format, otherwise basic format
+    Returns:
+        str : amount in dollar format
+    """
+    if report:  # XXXX.XX
+        return '{:.2f}'.format(amount)
+    else:  # X,XXX or X,XXX.XX
+        dollars = get_dollars(amount // 1)
+        cents = get_cents(amount % 1)
+        return dollars + cents
+
+
+def get_dollars(dollars):
+    """ Returns dollar amount as a string with 100s
```

Comment:
nice job -- but there is a format specifier for that:

```
In [13]: "${:,.2f}".format(1000000000)
Out[13]: '$1,000,000,000.00'
```
keep in mind that if there is something lots of people are going to need -- there is a good chance there is a built-in way to do it :-)

But I hope you had fun writing this!